content/issue-4/how-tokio-schedule-task.md

```diff
@@ -0,0 +1,115 @@
+`Future`, representing an asynchronous computation task, are the basis for implementing asynchronous in Rust. Unlike other languages, the computation does not automatically execute in the background, it needs to actively call its `poll` method. Tokio is the most widely used asynchronous runtime in the community. It uses various tips internally to ensure that `Future` are scheduled and executed fairly and timely. However, since the execution of Futures is cooperative, starvation of `Future` can inevitably occur in some scenarios.
+
+This article will analyze a problem I encountered while developing CeresDB to look at issues that may arise with Tokio scheduling. Please point out any inadequacies as my knowledge is limited.
+
+# Background
+[CeresDB](https://github.com/CeresDB/ceresdb) is a high-performance time series database designed for cloud-native. The storage engine uses an LSM-like architecture where data is first written to memtable, and when some threshold is reached, it is flushed to the underlying storage(e.g. S3). To prevent too many small files, there is also a background thread that does compaction.
+
+In production, I found a strange problem. Whenever compaction requests increased for a table, the flush time would spike even though flush and compaction run in different thread pools and have no direct relationship. Why did they affect each other?
+
+# Analysis
+To investigate the root cause, we need to understand Tokio's task scheduling mechanism. Tokio is event-driven, users submit tasks via `spawn`, then Tokio's scheduler decides how to execute them, most of time using a multi-threaded scheduler.
+
+Multi-threaded scheduler dispatches tasks to a fixed thread pool, each worker thread has a local run queue to save pending tasks. When starts, each worker thread will enter a loop to sequentially fetch and execute tasks in its run queue. Without a strategy, this scheduling can easily become imbalanced. Tokio uses work stealing to address this - when a worker's run queue is empty, it tries to "steal" tasks from other workers' queues to execute.
+
+In the above, a task is the minimum scheduling unit, corresponding to `await` in our code. Tokio can only reschedule on reaching an await, since future execution is a state machine like below:
```

Comment:
```suggestion
In the above, a task is the minimum scheduling unit, corresponding to `await` in our code. Tokio can only reschedule on reaching an await, since `Future` execution is a state machine like below:
```